ARCs/arc-0020.md

```diff
@@ -0,0 +1,869 @@
+>   defined in the [metadata section](#metadata). In addition:
+>   - The `manager`, `reserve` and `freeze` addresses SHOULD be set to the
+>     account of the controlling Smart Contract.
+>   - The remaining fields are left to the implementation, which MAY set `total`
+>     to `2 ** 64 - 1` to enable dynamically increasing the max circulating
+>     supply of the Smart ASA.
+>   - `name` and `unit_name` MAY be set to `SMART-ASA` and `S-ASA`, to
+>     denote that this ASA is Smart and has a controlling application.
+> - Persist the `total`, `decimals`, `default_frozen`, etc. fields for later
+>   use/retrieval.
+> - Return the ID of the created ASA.
+>
+> It is RECOMMENDED for calls to this method to be permissioned, e.g. to only
+> approve transactions issued by the controlling Smart Contract creator.
+
+#### Asset Configuration
+
+```json
+{
+  "name": "asset_config",
+  "args": [
+    { "type": "asset", "name": "config_asset" },
+    { "type": "uint64", "name": "total" },
+    { "type": "uint32", "name": "decimals" },
+    { "type": "bool", "name": "default_frozen" },
+    { "type": "string", "name": "unit_name" },
+    { "type": "string", "name": "name" },
+    { "type": "string", "name": "url" },
+    { "type": "byte[]", "name": "metadata_hash" },
+    { "type": "address", "name": "manager_addr" },
+    { "type": "address", "name": "reserve_addr" },
+    { "type": "address", "name": "freeze_addr" },
+    { "type": "address", "name": "clawback_addr" }
+  ],
+  "returns": { "type": "void" }
+}
+```
+
+Calling `asset_config` configures an existing Smart ASA.
+
+> Upon a call to `asset_config`, a reference implementation SHOULD:
+>
+> - Fail if `config_asset` does not correspond to an ASA controlled by this smart
+>   contract.
+> - Succeed iff the `sender` of the transaction corresponds to the `manager_addr`
+>   that was previously persisted for `config_asset` by a previous call to this
+>   method or, if never caller, to `asset_create`.
+> - Update the persisted `total`, `decimals`, `default_frozen`, etc. fields for later
+>   use/retrieval.
+>
+> The business logic associated to the update of the other parameters is left to
+> the implementation. An implementation that maximizes similarities with ASAs,
+> SHOULD NOT allow modifying the `clawback_addr` or `freeze_addr` after they
+> have been set to the special value `ZeroAddress`.
+>
+> The implementation MAY provide flexibility on the fields of an ASA that
+> cannot be updated after initial configuration. For instance, it MAY update the
+> `total` parameter to enable minting of new units or restricting the maximum
+> supply; when doing so, the implementation SHOULD ensure that the updated
+> `total` is not lower than the current circulating supply of the asset.
+
+#### Asset Transfer
+
+```json
+{
+  "name": "asset_transfer",
+  "args": [
+    { "type": "asset", "name": "xfer_asset" },
+    { "type": "uint64", "name": "asset_amount" },
+    { "type": "account", "name": "asset_sender" },
```

Comment:
The `asset_sender` field of the protocols `AssetTransferTransaction` is only for clawbacks, preferring `sender` for sending the asset. Should we keep the same naming convention here?

Reply:
I _think_ here we want this semantic:

- `sender` is always the user triggering the call, i.e. issuing the transaction
- Assuming permissions and roles, the call moves asset from A (`asset_sender`) to B (`asset_receiver`)
- `asset_sender`  COULD be `sender` and will probably be in most calls

Reply:
I agree. The `sender` is by default the issuer of the ABI call transaction. If `sender` is different from `asset_sender` (which should always be specified since ABI standard doesn't have optional arguments) the this is `asset_transfer` is interpreted as `clawback` and the consequent checks must follow.